### PR TITLE
add_complaint

### PR DESCRIPTION
**Pull request by KunzaadaBekzhanova**

### CODE REVIEW

**Review by narynbaev-seiitbek — APPROVED** [review]

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("api/complaints")
+@CrossOrigin
+
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:19]: Unnecessary line

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+@Tag(name = "Complaint API", description = "user with role \" User \"  can create send complaints, but cannot see all complaints\n Admin can only see all complaints")
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:20]: user **U** should be a capital letter

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+public class ComplaintApi {
+
+    private final ComplaintServiceImpl complaintService;
+
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "send complaint to Wish", description = "The user can send complain to Admin.")
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:26]: send S should be a capital

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @PostMapping("wish/{wishId}")
+    public SimpleResponse sendComplaintToWish(@PathVariable Long wishId, @RequestParam String text) {
+        return complaintService.sendComplaintToWish(wishId, text);
+    }
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @Operation(summary = "All complains", description = "Only Admin can see all complains")
+    @GetMapping("all")
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:33]: remove all

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    private final ComplaintServiceImpl complaintService;
+
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "send complaint to Wish", description = "The user can send complain to Admin.")
+    @PostMapping("wish/{wishId}")
+    public SimpleResponse sendComplaintToWish(@PathVariable Long wishId, @RequestParam String text) {
+        return complaintService.sendComplaintToWish(wishId, text);
+    }
+    @PreAuthorize("hasAuthority('ADMIN')")
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:31]: Next line

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    @Operation(summary = "All complains", description = "Only Admin can see all complains")
+    @GetMapping("all")
+    public List<ComplaintResponse> getAllComplaints(){
+        return complaintService.getComplaints();
+    }
+
+    @PreAuthorize("hasAuthority('USER')")
+    @Operation(summary = "send complaint to Gift", description = "The user can send complaint to Admin.")
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:39]: S capital

File: src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import kg.giftlist.giftlist.db.service.impl.ComplaintServiceImpl;
+import kg.giftlist.giftlist.dto.SimpleResponse;
+import kg.giftlist.giftlist.dto.mapper.complaint.ComplaintResponse;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/apis/ComplaintApi.java:14]: Pls refactor your code before pushing with **ctrl + alt + L**

File: src/main/java/kg/giftlist/giftlist/db/models/Complaint.java [hunk 8]
```diff
@@ -21,5 +21,15 @@ public class Complaint {
     private String text;
 
     @OneToOne
-    private User user;
+    @JoinColumn(name = "from_user_id")
+    private User fromUser;
+
+    @OneToOne
+    @JoinColumn(name = "wish_id")
+
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/db/models/Complaint.java:29]: Unnecessary line

File: src/main/java/kg/giftlist/giftlist/db/service/impl/ComplaintServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.ws.rs.ForbiddenException;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class ComplaintServiceImpl {
+    private final WishRepository wishRepository;
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/db/service/impl/ComplaintServiceImpl.java:29]: Next line

File: src/main/java/kg/giftlist/giftlist/db/service/impl/ComplaintServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
+package kg.giftlist.giftlist.db.service.impl;
+
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/db/service/impl/ComplaintServiceImpl.java:2]: Pls refactor class there is some unnecessary lines

File: src/main/java/kg/giftlist/giftlist/dto/mapper/complaint/ComplaintResponse.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.SequenceGenerator;
+
+@Getter
+@Setter
+@RequiredArgsConstructor
+public class ComplaintResponse {
+    private Long id;
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/dto/mapper/complaint/ComplaintResponse.java:22]: Next line

File: src/main/java/kg/giftlist/giftlist/dto/mapper/complaint/ComplaintResponse.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
```
Comment (peaksoft-school) [inline src/main/java/kg/giftlist/giftlist/dto/mapper/complaint/ComplaintResponse.java:23]: Unnecessary lines in this class

**Review by peaksoft-school — COMMENTED** [review]
Commented 21/08/22 at 2 pm

**Pull request merged**